### PR TITLE
[AMD][ROCm]Quantization methods on ROCm; Fix _scaled_mm call

### PR DESCRIPTION
**Pull request by gshtras**
Enabling compressed-tensors and fbgemm quantization on rocm in addition to fp8.
Fixing scaled_mm required parameters as scale_a and scale_b are [going to become](https://github.com/pytorch/pytorch/commit/fc2913fb808dd67667c4c57d01983a4dccec0f66#diff-e8a569efee1e650172f120a0fdcda024fe3e4703a4ee3336425c8f685af6b3abR890-R893) non-optional in 2.5rc1

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Review by mgoin — COMMENTED** [review]
Thanks for this PR! Do you support per-channel/per-token scales with this? I was under the assumption that only per-tensor scales were supported at the moment. If this is true, we should restrict support in the supported schemes for rocm, and I'm not sure how it works for fbgemm since they rely on dynamic per-token activation scales

**Comment by gshtras** [comment]
> Thanks for this PR! Do you support per-channel/per-token scales with this? I was under the assumption that only per-tensor scales were supported at the moment. If this is true, we should restrict support in the supported schemes for rocm, and I'm not sure how it works for fbgemm since they rely on dynamic per-token activation scales

The only difference on ROCm is the use of e4m3fnuz instead of e4m3fn, all the rest should work just the same, including the implementation of _scaled_mm, verified this on models quantized with fbgemm (meta llama 405b) and compressed-tensors.

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py [hunk 1]
```diff
@@ -39,16 +41,37 @@ def process_weights_after_loading(self, layer) -> None:
...
+                                                  requires_grad=False)
+
             layer.weight = Parameter(weight.t(), requires_grad=False)
             layer.weight_scale = Parameter(max_w_scale, requires_grad=False)
 
         # If channelwise, scales are already lined up, so just transpose.
         elif self.strategy == QuantizationStrategy.CHANNEL:
             weight = layer.weight
+
+            if is_hip():
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py:60]: can we move the is_hip() to a utils file? seem to be similar code in multiple cases

**Review by mgoin — COMMENTED** [review]
In that case, could we enable AMD hardware to run the existing fp8 tests?

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py (hunk 1 above)
Comment (gshtras) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py:60]: We could move the input_scale adjustment into utils, as it is the one common part between the 3 quantization types that support HIP fnuz), but IMO that would make it less explicit, because different parameters of the layer (weight, weight_scale, input_scale) would be adjusted in different places in the code.
Other than that it's a single function call from utils that's specific to HIP.

**Review by mgoin — COMMENTED** [review]
Aside from kernel tests, I am most interested in seeing E2E model loading and evaluation tests for some existing compressed tensors checkpoints. I think you could do this be making a new `vllm/.buildkite/lm-eval-harness/configs/models-amd.txt` config.

For instance we could define the list of models to just have FP8 models for now:
```
Meta-Llama-3-8B-Instruct.yaml
Meta-Llama-3-8B-Instruct-FP8-compressed-tensors.yaml
Minitron-4B-Base-FP8.yaml
```

See the existing lm-eval test here, so you could make a new entry for AMD hardware with the new set of models to test 
https://github.com/vllm-project/vllm/blob/0a4806f0a99880df1f74b10a6dceaf638cd3981c/.buildkite/test-pipeline.yaml#L270-L278

**Comment by gshtras** [comment]
> Aside from kernel tests, I am most interested in seeing E2E model loading and evaluation tests for some existing compressed tensors checkpoints. I think you could do this be making a new `vllm/.buildkite/lm-eval-harness/configs/models-amd.txt` config.
> 
> For instance we could define the list of models to just have FP8 models for now:
> 
> ```
> Meta-Llama-3-8B-Instruct.yaml
> Meta-Llama-3-8B-Instruct-FP8-compressed-tensors.yaml
> Minitron-4B-Base-FP8.yaml
> ```
> 
> See the existing lm-eval test here, so you could make a new entry for AMD hardware with the new set of models to test
> 
> https://github.com/vllm-project/vllm/blob/0a4806f0a99880df1f74b10a6dceaf638cd3981c/.buildkite/test-pipeline.yaml#L270-L278

That wouldn't be possible in the CI just yet. We don't have fp8 supporting hardware (MI300) there. What we could do is prepare these tests to be run manually, but not include in the CI

**Review by mgoin — APPROVED** [review]
Okay I see, I will approve this pending green CI then

**Comment by gshtras** [comment]
Some lm_eval numbers on the new quantization methods:
meta-llama/Meta-Llama-3.1-405B-FP8 (fbgemm_fp8)
|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|-----:|---|-----:|
|lambada_openai|      1|none  |     0|acc       |↑  |0.8166|±  |0.0054|
|              |       |none  |     0|perplexity|↓  |2.3953|±  |0.0395|

neuralmmagic/Meta-Llama-3.1-70B-Instruct-FP8 (compressed-tensors)
|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|-----:|---|-----:|
|lambada_openai|      1|none  |     0|acc       |↑  |0.7396|±  |0.0061|
|              |       |none  |     0|perplexity|↓  |4.1083|±  |0.1444|

Both with `lm_eval --model vllm --model_args pretrained=...,tensor_parallel_size=4 --tasks lambada_openai --batch_size 8`
